lib/pam/pam.c

```diff
@@ -0,0 +1,203 @@
+// +build pam,cgo
+
+#include "_cgo_export.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <dlfcn.h>
+#include <security/pam_appl.h>
+
+// library_name returns the name of the library to load at runtime.
+char *library_name()
+{
+#ifdef __APPLE__
+    return "libpam.dylib";
+#else
+    return "libpam.so";
+#endif
+}
+
+// converse is called by PAM to interact with the user. Interaction means
+// either writing something to stdout and stderr or reading something from
+// stdin.
+int converse(int n, const struct pam_message **msg, struct pam_response **resp, void *data)
+{
+    int i;
+    struct pam_response *aresp;
+
+    // If no messages arrived, or the number of messages is greater than
+    // allowed, something is wrong with the caller.
+    if (n <= 0 || n > PAM_MAX_NUM_MSG) {
+        return PAM_CONV_ERR;
+    }
+
+    // According to pam_conv(3): "It is the caller's responsibility to release
+    // both, this array and the responses themselves, using free(3)." The
+    // caller in this situation is the PAM module and the array and the
+    // responses refer to aresp and aresp[i].resp.
+    aresp = calloc(n, sizeof *aresp);
+    if (aresp == NULL) {
+        return PAM_BUF_ERR;
+    }
+
+    // Loop over all messages and process them.
+    for (i = 0; i < n; ++i) {
+        aresp[i].resp_retcode = 0;
+        aresp[i].resp = NULL;
+
+        switch (msg[i]->msg_style) {
+        case PAM_PROMPT_ECHO_OFF:
+            // Read back response from user. What the user writes should not
+            // be echoed to the screen.
+            aresp[i].resp = readCallback((uintptr_t)data, 0);
+            if (aresp[i].resp == NULL) {
+                goto fail;
+            }
+            break;
+        case PAM_PROMPT_ECHO_ON:
+            // First write the message to stderr.
+            writeCallback((uintptr_t)data, STDERR_FILENO, (char *)(msg[i]->msg));
+
+            // Read back response from user. What the user writes will be
+            // echoed to the screen.
+            aresp[i].resp = readCallback((uintptr_t)data, 1);
+            if (aresp[i].resp == NULL) {
+                goto fail;
+            }
+            break;
+        case PAM_ERROR_MSG:
+            // Write message to stderr.
+            writeCallback((uintptr_t)data, STDERR_FILENO, (char *)(msg[i]->msg));
+            if (strlen(msg[i]->msg) > 0 && msg[i]->msg[strlen(msg[i]->msg) - 1] != '\n') {
```

Comment:
not sure I get this check logic, can you elaborate? I get the length check part, but why \n at the end is special?